Auth: Duplicate API Key Name Handle With Useful HTTP Code

ATTron opened the pull request:
**What this PR does / why we need it**:
adds useful http status code & message when trying to create api key 
with same name as another existing api key
**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #17447

ATTron committed:
API: Duplicate API Key Name Handle With Useful HTTP Code

papagian reviewed (changes requested):
Great, thank you for your contribution!
I have some suggestions before merging it.

[Here](https://github.com/ATTron/grafana/blob/17447-Duplicate-API-Key-Name-Handling/pkg/services/sqlstore/apikey.go#L59) you could make use `dialect.IsUniqueConstraintViolation` and return your error `ErrDuplicateApiKey` just as it happens [here](https://github.com/ATTron/grafana/blob/17447-Duplicate-API-Key-Name-Handling/pkg/services/sqlstore/apikey.go#L46)

papagian commented on pkg/api/apikey.go:
```diff
@@ -68,7 +68,10 @@ func (hs *HTTPServer) AddAPIKey(c *models.ReqContext, cmd models.AddApiKeyComman
 			return Error(400, err.Error(), nil)
 		}
-		return Error(500, "Failed to add API key", err)
+		if err.Error() == models.ErrDuplicateApiKey.Error() {
```
Following my suggestion for using `dialect.IsUniqueConstraintViolation` and returning the actual error this can be changed to:
`if err == models.ErrDuplicateApiKey {`

papagian commented on pkg/api/apikey.go:
```diff
@@ -68,7 +68,10 @@ func (hs *HTTPServer) AddAPIKey(c *models.ReqContext, cmd models.AddApiKeyComman
 		}
-		return Error(500, "Failed to add API key", err)
+		if err.Error() == models.ErrDuplicateApiKey.Error() {
+			return Error(409, "API Key Name Must Be Unique", nil)
```
And this can be changed to:
`return Error(409, err.Error(), nil)`
just as it happens for `ErrInvalidApiKeyExpiration` just above

papagian commented on pkg/models/apikey.go:
```diff
@@ -7,6 +7,7 @@ import (
 
 var ErrInvalidApiKey = errors.New("Invalid API Key")
 var ErrInvalidApiKeyExpiration = errors.New("Negative value for SecondsToLive")
+var ErrDuplicateApiKey = errors.New("UNIQUE constraint failed: api_key.org_id, api_key.name")
```
Finally, this can become:
`var ErrDuplicateApiKey = errors.New("API Organisation id and Key Name Must Be Unique")`

ATTron committed:
Revert "17447: make changes requested during review"

This reverts commit a4a674ea83a9288701611f203f2a75531fb8a131.

ATTron committed:
17447: make changes requested during review

- use dialect.IsUniqueContraintViolation
- change if statement to match others
- return error properly

ATTron commented:
I've made the changes you've suggested. Let me know if you need anything else from me!

papagian commented:
Sorry, your implementation is absolutely fine but after reconsidering it a bit I think I would like this to be implemented in a different way because usage of `dialect.IsUniqueConstraintViolation` should be generally avoided and only used in specific scenarios.

So, I would suggest you to revert the last commit and instead of handling the unique constrain error, a check for existing key with the specific organisation id and name is performed before the insert.
In that case if the organisation id and name combination is reserved the error `models.ErrDuplicateApiKey` will be returned (similarly to [this](https://github.com/grafana/grafana/blob/78ca55f3d73019f2e942e8cec2987c32274e3f5f/pkg/services/sqlstore/datasource.go#L93-L98) code).

Also, it's not obligatory but it would be great if you have some more time to invest to write a test similar to these [tests](https://github.com/grafana/grafana/blob/master/pkg/services/sqlstore/apikey_test.go#L10) that tries to insert an api key with a reserved organisation and name and gets the expected error.

Sorry again for changing my mind! Please let me know if you are you still interested to make these modifications?

ATTron committed:
API: useful http code on duplicate api key error w/ tests

ATTron commented:
>  Sorry again for changing my mind! Please let me know if you are you still interested to make these modifications?

Not a problem at all! I pushed what you suggested and added testing as well.

papagian commented on pkg/api/apikey.go:
```diff
@@ -68,7 +68,10 @@ func (hs *HTTPServer) AddAPIKey(c *models.ReqContext, cmd models.AddApiKeyComman
+		if err == models.ErrDuplicateApiKey {
+			return Error(409, err.Error(), nil)
+		}
+		return Error(500, "Failed to add API", err)
```
This error message used to "Failed to add API key"

papagian reviewed (changes requested):
Great, thanks!
Almost there, just one more small change.

ATTron committed:
API: API Key Duplicate Handling

fixed small typo associated with error

ATTron commented:
All good to go

papagian commented:
Thank you for contributing to Grafana!
